firedrake/preconditioners/pmg.py

```diff
@@ -507,100 +523,169 @@ def prolongation_transfer_kernel_action(Vf, expr):
-    elif ele.cell() == ufl.hexahedron:
-        hexahedron_tpc = ufl.TensorProductCell(ufl.quadrilateral, ufl.interval)
-        return expand_element(ele.reconstruct(cell=hexahedron_tpc))
-    elif isinstance(ele, (ufl.TensorElement, ufl.VectorElement)):
-        return expand_element(ele._sub_element)
-    elif isinstance(ele, (ufl.HDivElement, ufl.HCurlElement, ufl.BrokenElement, ufl.RestrictedElement)):
-        return expand_element(ele._element)
-    elif isinstance(ele, ufl.WithMapping):
-        return expand_element(ele.wrapee)
-    elif isinstance(ele, ufl.MixedElement):
-        return ufl.MixedElement(*[expand_element(e) for e in ele.sub_elements()])
-    elif isinstance(ele, ufl.EnrichedElement):
-        terms = []
-        for e in ele._elements:
-            ee = expand_element(e)
-            if isinstance(ee, ufl.EnrichedElement):
-                terms.extend(ee._elements)
-            else:
-                terms.append(ee)
-        return ufl.EnrichedElement(*terms)
-    elif isinstance(ele, ufl.TensorProductElement):
-        factors = [expand_element(e) for e in ele.sub_elements()]
+    # Expand a FiniteElement as an EnrichedElement of TensorProductElements, discarding modifiers.
+    if isinstance(ele, finat.FlattenedDimensions):
+        return expand_element(ele.product)
+    elif isinstance(ele, (finat.HDivElement, finat.HCurlElement)):
+        return expand_element(ele.wrappee)
+    elif isinstance(ele, finat.DiscontinuousElement):
+        return expand_element(ele.element)
+    elif isinstance(ele, finat.EnrichedElement):
+        terms = list(map(expand_element, ele.elements))
+        return finat.EnrichedElement(terms)
+    elif isinstance(ele, finat.TensorProductElement):
+        factors = list(map(expand_element, ele.factors))
         terms = [tuple()]
         for e in factors:
             new_terms = []
-            for f in e._elements if isinstance(e, ufl.EnrichedElement) else [e]:
-                f_factors = f.sub_elements() if isinstance(f, ufl.TensorProductElement) else (f,)
+            for f in e.elements if isinstance(e, finat.EnrichedElement) else [e]:
+                f_factors = tuple(f.factors) if isinstance(f, finat.TensorProductElement) else (f,)
                 new_terms.extend([t_factors + f_factors for t_factors in terms])
             terms = new_terms
-        if len(terms) == 1:
-            return ufl.TensorProductElement(*terms[0])
-        else:
-            return ufl.EnrichedElement(*[ufl.TensorProductElement(*k) for k in terms])
+        terms = list(map(finat.TensorProductElement, terms))
+        return finat.EnrichedElement(terms)
     else:
         return ele
 
 
-def get_line_elements(V):
+def evaluate_dual(dual, element, key=None):
+    # Evaluate the action of a set of dual functionals on the basis functions of an element.
+    keys = set(tuple(phi.get_point_dict().keys()) for phi in dual)
+    pts = list(set(sum(keys, ())))
```

Comment:
These are not deterministically ordered, does it matter?

Reply:
I believe that since we are integrating by summing over quadrature points the order of summation might have an effect on the result, causing different kernel code to be compiled on different processes. I guess `numpy.unique` will DTRT